relay/polkadot/src/coretime_migration.rs

```diff
@@ -0,0 +1,319 @@
+			CoreIndex(core as u32),
+			now,
+			vec![(CoreAssignment::Task(para_id.into()), PartsOf57600::FULL)],
+			None,
+		);
+		if let Err(err) = r {
+			log::error!(
+				"Creating assignment for existing para failed: {:?}, error: {:?}",
+				para_id,
+				err
+			);
+		}
+	}
+
+	let config = configuration::ActiveConfig::<T>::get();
+	// num_cores was on_demand_cores until now:
+	for on_demand in 0..config.scheduler_params.num_cores {
+		let core = CoreIndex(legacy_count.saturating_add(on_demand as _));
+		let r = parachains_assigner_coretime::Pallet::<T>::assign_core(
+			core,
+			now,
+			vec![(CoreAssignment::Pool, PartsOf57600::FULL)],
+			None,
+		);
+		if let Err(err) = r {
+			log::error!("Creating assignment for existing on-demand core, failed: {:?}", err);
+		}
+	}
+	let total_cores = config.scheduler_params.num_cores + legacy_count;
+	configuration::ActiveConfig::<T>::mutate(|c| {
+		c.scheduler_params.num_cores = total_cores;
+	});
+
+	if let Err(err) = migrate_send_assignments_to_coretime_chain::<T, SendXcm, LegacyLease>() {
+		log::error!("Sending legacy chain data to coretime chain failed: {:?}", err);
+	}
+
+	let single_weight = <T as Config>::WeightInfo::assign_core(1);
+	single_weight
+		.saturating_mul(u64::from(legacy_count.saturating_add(config.scheduler_params.num_cores)))
+		// Second read from sending assignments to the coretime chain.
+		.saturating_add(T::DbWeight::get().reads_writes(2, 1))
+}
+
+fn migrate_send_assignments_to_coretime_chain<
+	T: Config,
+	SendXcm: xcm::v4::SendXcm,
+	LegacyLease: GetLegacyLease<BlockNumberFor<T>>,
+>() -> result::Result<(), SendError> {
+	let legacy_paras = LegacyLease::get_all_parachains_with_leases();
+	let legacy_paras_count = legacy_paras.len();
+	let (system_chains, lease_holding): (Vec<_>, Vec<_>) =
+		legacy_paras.into_iter().partition(IsSystem::is_system);
+
+	let reservations = system_chains.into_iter().map(|p| {
+		let schedule = BoundedVec::truncate_from(vec![ScheduleItem {
+			mask: CoreMask::complete(),
+			assignment: CoreAssignment::Task(p.into()),
+		}]);
+		mk_coretime_call::<T>(CoretimeCalls::Reserve(schedule))
+	});
+
+	let mut leases = lease_holding.into_iter().filter_map(|p| {
+			log::trace!(target: "coretime-migration", "Preparing sending of lease holding para {:?}", p);
+			let Some(valid_until) = LegacyLease::get_parachain_lease_in_blocks(p) else {
+				log::error!("Lease holding chain with no lease information?!");
+				return None
+			};
+			let valid_until: u32 = match valid_until.try_into() {
+				Ok(val) => val,
+				Err(_) => {
+					log::error!("Converting block number to u32 failed!");
+					return None
+				},
+			};
+			// We assume the coretime chain set this parameter to the recommended value in RFC-1:
+			const TIME_SLICE_PERIOD: u32 = 80;
+			let round_up = if valid_until % TIME_SLICE_PERIOD > 0 { 1 } else { 0 };
+			let time_slice = valid_until / TIME_SLICE_PERIOD + TIME_SLICE_PERIOD * round_up;
```

Comment:
```suggestion
			let time_slice = (valid_until + TIME_SLICE_PERID - 1) / TIME_SLICE_PERIOD;
```

Reply:
I think we do a different kind of round up here. `timeslice` is not number of periods but a lifetime in blocks (divided by 80). Your formula works only for the 'number of periods' case.

Reply:
I'm confused... why we multiply the round up with `TIME_SLICE_PERIOD`?

Shouldn't it be `let time_slice = valid_until / TIME_SLICE_PERIOD + round_up;`) which is equivalent to your formula?

Reply:
`round_up` is just `0` or `1`. It leads to the same result as what I have written there.

Reply:
It doesn't. Round up is multiplied by TIME_SLICE_PERIOD. Have a look: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=a5ce70639f2ee8a169e5d618e707ed74

Reply:
Fixed